assembly/src/tokens/tokenizer.rs

```diff
@@ -0,0 +1,169 @@
+use super::{LineInfo, SourceLocation, Token};
+use core::mem;
+
+// LINE TOKENIZER
+// ================================================================================================
+
+/// A line tokenizer that will generate tokens with their locations from [LinesStream] and
+/// [LineInfo].
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct LineTokenizer<'a> {
+    line: &'a str,
+    location: SourceLocation,
+    dangling: Option<SourceLocation>,
+}
+
+impl<'a> From<&LineInfo<'a>> for LineTokenizer<'a> {
+    fn from(info: &LineInfo<'a>) -> Self {
+        let line_number = info.line_number();
+        let column = info.char_offset() + 1;
+        let location = SourceLocation::new(line_number, column);
+        let line = info.contents().unwrap_or("");
```

Comment:
`LineTokenizer` should assume that a line starts with a valid token - so, we should replace `unwrap_or()` with `expect()`.

Also, we could add `debug_assert` here to make sure that line content doesn't start with an a whitespace or a comment.

Reply:
I think returning `Option` and then doing `expect()` is fine. I would have probably just done `debug_assert` since the only way this would fail is if we have a bug in `TokenStream` - but your way is safer.